source/common/http/header_utility.h

```diff
@@ -65,30 +67,195 @@ class HeaderUtility {
+      // treat the header value as empty.
+      const auto value = header_value.result().has_value() ? header_value.result().value() : "";
+      // Execute the specific matcher's code and invert if invert_match_ is set.
+      return specificMatchesHeaders(value) != invert_match_;
+    };
 
+  protected:
+    // A matcher specific implementation to match the given header_value.
+    virtual bool specificMatchesHeaders(absl::string_view header_value) const PURE;
     const LowerCaseString name_;
-    HeaderMatchType header_match_type_;
-    std::string value_;
-    Regex::CompiledMatcherPtr regex_;
-    envoy::type::v3::Int64Range range_;
-    Matchers::StringMatcherPtr string_match_;
     const bool invert_match_;
     const bool treat_missing_as_empty_;
-    bool present_;
+  };
 
-    // HeaderMatcher
-    bool matchesHeaders(const HeaderMap& headers) const override {
-      return HeaderUtility::matchHeaders(headers, *this);
+  // Corresponds to the exact_match from the HeaderMatchSpecifier proto in the RDS API.
+  class HeaderDataExactMatch : public HeaderDataBaseImpl {
+  public:
+    HeaderDataExactMatch(const envoy::config::route::v3::HeaderMatcher& config)
+        : HeaderDataBaseImpl(config), expected_value_(config.exact_match()) {}
+
+  private:
+    bool specificMatchesHeaders(absl::string_view header_value) const override {
+      return expected_value_.empty() || header_value == expected_value_;
+    };
+    const std::string expected_value_;
+  };
+
+  // Corresponds to the safe_regex_match from the HeaderMatchSpecifier proto in the RDS API.
+  class HeaderDataRegexMatch : public HeaderDataBaseImpl {
+  public:
+    HeaderDataRegexMatch(const envoy::config::route::v3::HeaderMatcher& config,
+                         Server::Configuration::CommonFactoryContext& factory_context)
+        : HeaderDataBaseImpl(config),
+          regex_(Regex::Utility::parseRegex(config.safe_regex_match(),
+                                            factory_context.regexEngine())) {}
+
+  private:
+    bool specificMatchesHeaders(absl::string_view header_value) const override {
+      return regex_->match(header_value);
+    };
+    const Regex::CompiledMatcherPtr regex_;
+  };
+
+  // Corresponds to the range_match from the HeaderMatchSpecifier proto in the RDS API.
+  class HeaderDataRangeMatch : public HeaderDataBaseImpl {
+  public:
+    HeaderDataRangeMatch(const envoy::config::route::v3::HeaderMatcher& config)
+        : HeaderDataBaseImpl(config) {
+      range_.set_start(config.range_match().start());
+      range_.set_end(config.range_match().end());
+    }
+
+  private:
+    bool specificMatchesHeaders(absl::string_view header_value) const override {
+      int64_t header_int_value = 0;
+      return absl::SimpleAtoi(header_value, &header_int_value) &&
+             header_int_value >= range_.start() && header_int_value < range_.end();
+    };
+
+    // TODO(adisuissa): convert from protobuf to two ints.
```

Comment:
Might as well fix this now while you're touching this...

Reply:
Ok, initially I thought we should just make this PR compatible with how it previously was handling things, but I agree it is not a big change.
Done.